tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -699,3 +768,168 @@ def prepare_solution_vector(self, number_density):
+        t_electrons,
+        j_blues,
+        beta_sobolev,
+        excitation_species,
+        nlte_data,
+    ):
+        """Calculates a matrix with bound-bound rates for NLTE excitation treatment.
+
+        Parameters
+        ----------
+        atomic_data : Object
+            Atomic data from the atomic datafile.
+        t_electrons :  Numpy Array, dtype float
+            Electron temperatures.
+        j_blues : Pandas DataFrame, dtype float
+            J_blue values as calculated in LTE.
+        beta_sobolev : Numpy Array, dtype float
+            Sobolev escape probability
+        excitation_species : tuple
+            Species treated in NLTE excitation.
+        nlte_data : NLTEExcitationData
+            Data relevant to NLTE exciation species.
+
+        Returns
+        -------
+        numpy.array
+            Matrix with excitation-deexcitation rates(should be added to NLTE rate matrix for excitation treatment).
+        """
+
+        number_of_levels = atomic_data_levels.energy.loc[
+            excitation_species
+        ].count()
+        lnl = nlte_data.lines_level_number_lower[excitation_species]
+        lnu = nlte_data.lines_level_number_upper[excitation_species]
+        (lines_index,) = nlte_data.lines_idx[excitation_species]
+
+        try:
+            j_blues_filtered = j_blues.iloc[lines_index]
+        except AttributeError:
+            j_blues_filtered = j_blues
+        try:
+            beta_sobolev_filtered = beta_sobolev.iloc[lines_index]
+        except AttributeError:
+            beta_sobolev_filtered = beta_sobolev
+        A_uls = nlte_data.A_uls[excitation_species]
+        B_uls = nlte_data.B_uls[excitation_species]
+        B_lus = nlte_data.B_lus[excitation_species]
+        r_lu_index = lnu * number_of_levels + lnl
+        r_ul_index = lnl * number_of_levels + lnu
+        r_ul_matrix = np.zeros(
+            (number_of_levels, number_of_levels, len(t_electrons)),
+            dtype=np.float64,
+        )
+        r_ul_matrix_reshaped = r_ul_matrix.reshape(
+            (number_of_levels**2, len(t_electrons))
+        )
+        r_ul_matrix_reshaped[r_ul_index] = (
+            A_uls[np.newaxis].T + B_uls[np.newaxis].T * j_blues_filtered
+        )
+        r_ul_matrix_reshaped[r_ul_index] *= beta_sobolev_filtered
+        r_lu_matrix = np.zeros_like(r_ul_matrix)
+        r_lu_matrix_reshaped = r_lu_matrix.reshape(
+            (number_of_levels**2, len(t_electrons))
+        )
+        r_lu_matrix_reshaped[r_lu_index] = (
+            B_lus[np.newaxis].T * j_blues_filtered * beta_sobolev_filtered
+        )
+
+        rates_matrix_bound_bound = r_lu_matrix + r_ul_matrix
+        for i in range(number_of_levels):
+            rates_matrix_bound_bound[i, i] = -rates_matrix_bound_bound[
+                :, i
+            ].sum(axis=0)
+        return rates_matrix_bound_bound
+
+        # TODO: beta sobolev needs to be recalculated for each iteration, because it depends on numberdensity
```

Comment:
All of the stuff that uses the beta_sobolevs should be moved to a different method that only works on a single shell at a time. We want to solve the non-linear equations for each shell separately, which requires us to recalculate the bound-bound rate matrix with new values of beta_sobolev at every evaluation of the solver. The current method is still useful since we can pre-calculate a lot of the things we need.

Reply:
When calling this, I was thinking of calling it for the individual shell(as our current solver also works that way for NLTE ionization treatment). Should I still move the stuff connected to beta_sobolevs to another method?

Reply:
The way it is coded it is intended to be called with multiple shells, see, e.g.:
`r_lu_matrix_reshaped = r_lu_matrix.reshape((number_of_levels**2, len(t_electrons)))`
You can pass in `len(t_electrons) = 1` but that is not the best solution.

It is not crucial to move the stuff connected to beta_sobolevs to another method but it would save computational time since we can skip half of the steps in setting up the rate matrix (which we will need to do a lot).